rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+
+* basic confirmation (y/n) prompt in the key agent, similar to what we use for headless
+* password prompt in the dependent client
+  * like with ssh-agent, this would be a temporary password provided on agent
+  startup
+* touch prompt with temporary local MFA registration
+
+Once the connection is authenticated through the local confirmation mechanism,
+the dependent client could perform any queries or signatures until client is
+closed. In practice, this means that each `tsh` command would require a single
+confirmation prompt to complete.
+
+Note: once again, we run into the issue of prompting on every `tsh` command
+that we ran into with `hardware_key_pin`. Therefore we would only want to
+enable this confirmation layer when the tradeoff of better security for worse
+UX is justified. See the sections on [the risk of key agent forwarding](#key-agent-forwarding)
+and [safely supporting key agent forwarding](#support-teleport-key-agent-forwarding-future-work)
+for more details on how and why this confirmation layer would be included.
+
+### Modes
+
+The Teleport Key Agent can work in two different modes; `signing` and `login sync`.
+
+Note: the Teleport Key Agent will be pre-released with `signing` mode only as
+it will greatly improve the UX for Hardware Key pin support, especially when
+paired with [Hardware Key pin caching](./0198-hardware-key-pin-caching).
+
+#### Signing mode
+
+In signing mode, the Teleport Key Agent will only be responsible for providing
+a signing interface for Teleport Clients, similar to ssh-agent or gpg-agent.
+All other login information (profile info, certificates, etc.) will be retrieved
+from the standard file storage interface.
+
+This signing mode provides the baseline feature set needed to introduce UX
+improvements for features which limit access to private keys, including:
+
+* Hardware Key Support
+* [Memory Key Storage (*future work*)](#memory-key-storage)
+
+Since only private keys are synced between clients in this mode, there is a
+possibility of state drift for other login information.
+
+For example, if the user is running Teleport Connect as the agent, but they
+relogin with `tsh`, Teleport Connect would not know that the keys and certs
+in their shared home directory have been replaced. Depending on what Teleport
+Connect was trying to do at the time of the login, it might accidentally load
+the new certs with the old key. This would lead to a new confusing error and,
+in the best case, another prompt to re-login.
+
+Note: in signing mode, dependent clients can perform any login operations as
+usual (e.g. `tsh login`, `tsh app login`) adding or removing certs directly in
+file storage. The dependent client will not generate new private keys in file
+storage and instead continue to use the same private keys offered by the agent
+via its sign method. This will prevent the agent from becoming out of sync
+with dependent clients.
+
+#### Login sync mode (*future work*)
```

Comment:
If we punt this as future work I'm afraid it will never get done. Should we take the time to do it now, given that we are already committed to building so much of this?

Reply:
The timeline for this feature doesn't allow for such a complicated piece of work, nor is it necessary for the MVP for hardware key pin UX improvements. I do think it's one of the better ways we can implement [shared login state between `tsh` and Teleport Connect](https://github.com/gravitational/teleport/issues/25806), so that's when we'd get to this work.